setup/helpers/abs_helper.rb

```diff
@@ -655,23 +663,123 @@ def valid_abs_resource_hosts?(abs_resource_hosts)
     else
 
       begin
-        hosts = JSON.parse(abs_resource_hosts)
-        host = hosts[0]
-        hostname = host["hostname"]
+        hosts = parse_abs_resource_hosts(abs_resource_hosts)
+        hostname = hosts[0]["hostname"]
```

Comment:
The validity is confirmed by ensuring that the first entry has a hostname. This check could be moved into `parse_abs_resource_hosts` and cause a raise if not found which would simplify `valid_abs_resource_hosts?`. I added `parse_abs_resource_hosts` to avoid parsing the JSON in multiple places but wanted to maintain the current behavior to avoid further refactoring at this point. 

`valid_abs_resource_hosts?` was intended to verify the contents of the `last_abs_resource_hosts.log` file while in the case of `provision_host_for_role` it is parsing the output of `get_abs_resource_hosts` which is now being called with a single host per role rather than the initial 'Apples2Apples' design which passed the `master` and `metrics` hosts in the same array.

The whole thing needs a redesign at this point :)

Reply:
Sorry, I did not express myself very well.  I understand the validation that is particular line provides an the work done in the rescue block.  With those validations in place, the previously existing validations are no longer necessary.  The method can be refactored because either `hostname = hosts[0]["hostname"]` throws an error or is good:

```
def valid_abs_resource_hosts?(abs_resource_hosts)
  is_valid = true
  begin
    hosts = parse_abs_resource_hosts(abs_resource_hosts)
    hostname = hosts[0]["hostname"]
  rescue StandardError => e
    puts "#{e.inspect}  encountered parsing the hosts array: #{abs_resource_hosts}"
    is_valid = false
  end
  return is_valid
end
```

Reply:
I actually moved the hostname check to `parse_abs_resource_hosts` since parsing should fail if it doesn't exist. I wanted to differentiate between the 'invalid JSON' vs 'invalid abs_resource_hosts' cases and provide a corresponding error message, however I can consolidate if that seems unnecessary.

Reply:
It does seem unnecessary to me, but if that is the desire, that is not what is happening now.

If I call `valid_abs_resource_hosts?('Foobar')`. I have passed an _invalid_ JSON string, but I get back the `encountered parsing the hosts array` error.
The same error is returned if I pass a _valid_JSON string ('{}')
However, if I pass `nil`, then I get the error `A valid JSON string is required...`.

So, I am confused about the intent of the error cases these two methods are trying to capture.